kvserver: make lease preference enforcement more responsive

kvoli opened the pull request:
When a leaseholder is lost, any surviving replica may acquire the lease, even if it violates lease preferences. There are two main reasons for this: we need to elect a new Raft leader who will acquire the lease, which is agnostic to lease preferences, and there may not even be any surviving replicas that satisfy the lease preferences at all, so we don't want to keep the range unavailable while we try to figure this out (considering e.g. network timeouts can delay this for many seconds).

However, after acquiring a lease, we rely on the replicate queue to transfer the lease back to a replica that conforms with the preferences, which can take several minutes. In multi-region clusters, this can cause severe latency degradation if the lease is acquired in a remote region.

This patch will detect lease preference violations when a replica acquires a new lease, and eagerly enqueue it in the replicate queue for transfer (if possible).

Resolves https://github.com/cockroachdb/cockroach/issues/106100.
Epic: none

Release note (bug fix): When losing a leaseholder and using lease preferences, the lease can be acquired by any other replica (regardless of lease preferences) in order to restore availability as soon as possible. The new leaseholder will now immediately check if it violates the lease preferences, and attempt to transfer the lease to a replica that satisfies the preferences if possible.

kvoli committed:
kvserver: [dnm] deflake lease preferences during outage

`TestLeasePreferencesDuringOutage` checks that a non-preferred node can
up-replicate and transfer the lease in a single replicate queue
`process()`

It was possible that another node would acquire the lease, causing the
replicate queue enqueue to fail.

Disable expiration based lease transfers and force the transfer to a
non-preferred locality. This avoids flakiness in parts of the test which
are not directly part of the behavior being asserted on.

This patch also unskips the test.

Resolves: #105101

Release note: None

erikgrinaker committed:
constraint: tighten inputs to `CheckConjunction`

Previously, `ConjunctionsCheck` took a store descriptor as an input.
However, it only needed to know the store/node attributes and locality.
Some upcoming callers (lease acquisition) can't easily construct a full
store descriptor since the locking order would cause deadlocks.

This patch changes it to only take the attributes and locality instead
of the entire store descriptor, and renames it to `CheckConjunction()`.
It also adds `CheckStoreConjunction()` as a convenience method that
takes a store descriptor, and migrates all existing callers.

Epic: none
Release note: None

erikgrinaker committed:
kvserver: transfer lease when acquiring lease outside preferences

When a leaseholder is lost, any surviving replica may acquire the lease,
even if it violates lease preferences. There are two main reasons for
this: we need to elect a new Raft leader who will acquire the lease,
which is agnostic to lease preferences, and there may not even be any
surviving replicas that satisfy the lease preferences at all, so we
don't want to keep the range unavailable while we try to figure this out
(considering e.g. network timeouts can delay this for many seconds).

However, after acquiring a lease, we rely on the replicate queue to
transfer the lease back to a replica that conforms with the preferences,
which can take several minutes. In multi-region clusters, this can cause
severe latency degradation if the lease is acquired in a remote region.

This patch will detect lease preference violations when a replica
acquires a new lease, and eagerly enqueue it in the replicate queue for
transfer (if possible).

Epic: none
Release note (bug fix): When losing a leaseholder and using lease
preferences, the lease can be acquired by any other replica (regardless
of lease preferences) in order to restore availability as soon as
possible. The new leaseholder will now immediately check if it violates
the lease preferences, and attempt to transfer the lease to a replica
that satisfies the preferences if possible.

Co-authored-by: Erik Grinaker <grinaker@cockroachlabs.com>
Co-authored-by: Austen McClernon <austen@cockroachlabs.com>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107126)
<!-- Reviewable:end -->

kvoli committed:
roachtest: [dnm] add lease preference speed tests

wip

Epic: none

Release note: None

kvoli committed:
kvserver: requeue in repl queue on pref violation

wip

Epic: none

Release note: None

kvoli committed:
kvserver: [dnm] check lease after repl q process

wip

Epic: none

Release note: None

kvoli committed:
allocator/allocatorimpl: log v=5 follow-the-work..

`shouldTransferLeaseForAccessLocality` logs detailed information
regarding the replica locality, QPS and request weighting. This
information is not summarized. Bump to v=5, from v=1-v=2 to reduce
vmodule enabled log noise.

Epic: none

Release note: None

kvoli committed:
kvserver: add lease preference metrics

There were no existing metrics to monitor the lease preference
enforcement. This commit adds two metric gauges:
`leases.preferences.violating` and `leases.preferences.less-preferred`.
These metrics are reported by the store.

`leases.preferences.violating` indicates the number of valid leases a
store owns, which satisfy none of the preferences applied.

`leases.preferences.less-preferred` indicates the number of valid leases
a store owns, which satisfy some of the preferences applied, but not the
first one.

When no preferences are applied, the lease is not counted in either
metric.

Epic: none

Release note (ops change): TODO

kvoli committed:
allocator/plan: place leases violating preferences in purgatory

This patch places replicas in the replicate queue purgatory when
it has a lease violating the lease preferences and it's unable to find a
suitable target. This causes the replica to be retried more often.

This will only trigger when replicas are eagerly enqueued (typically
when we acquire a new lease that violates preferences), since we
otherwise don't attempt to enqueue replicas when they don't have a valid
lease transfer target.

Epic: none
Release note: None

Co-authored-by: Erik Grinaker <grinaker@cockroachlabs.com>
Co-authored-by: Austen McClernon <austen@cockroachlabs.com>

kvoli committed:
kvserver: rename repl q re-(processing/re-queuing)

This naming was incorrect after #85219. The replicate queue will
re-queue , rather than immediately re-process a replica after a
successful process.

Epic: none

Release note: None

kvoli commented:
Going to close this in favor of another testing branch, then push onto https://github.com/cockroachdb/cockroach/pull/106079 for review.

kvoli commented:
Going to close this in favor of another testing branch, then push onto https://github.com/cockroachdb/cockroach/pull/106079 for review.
